packages/flutter/lib/src/gestures/binding.dart

```diff
@@ -99,8 +109,34 @@ mixin GestureBinding on BindingBase implements HitTestable, HitTestDispatcher, H
 
   void _flushPointerEventQueue() {
     assert(!locked);
-    while (_pendingPointerEvents.isNotEmpty)
-      _handlePointerEvent(_pendingPointerEvents.removeFirst());
+
+    // Consume all pending pointer events.
+    while (_pendingPointerEvents.isNotEmpty) {
+      final PointerEvent event = _pendingPointerEvents.removeFirst();
+      // Add touch events to resampler if resampling is enabled,
+      // otherwise dispatch pointer event directly.
+      if (resamplingEnabled && event.kind == PointerDeviceKind.touch) {
+        final PointerEventResampler resampler = _resamplers.putIfAbsent(
+            event.device, () => PointerEventResampler());
+        resampler.addEvent(event);
+      } else {
+        _handlePointerEvent(event);
+      }
+    }
+
+    // Iterate over active resamplers and sample pointer events for
+    // current `_sampleTime`.
+    _resamplers.forEach((int key, PointerEventResampler resampler) {
+      resampler.sample(_sampleTime, _handlePointerEvent);
```

Comment:
Isn't it possible that _sampleTime is outdated when this code runs? For example, after some frames have been rendered, the first input event is received. At that point, this code runs, but _sampleTime is still zero, no?

Reply:
Yes, but that just results in us dispatching the first event directly, which is fine in practice as that's typically a PointerAddedEvent. However, this has improved in latest change as we're now using the current scheduler frame time and the offset directly.